lib/logstash/filters/multiline.rb

```diff
@@ -246,35 +243,66 @@ def previous_filter!(event, match)
 
     if match
+      # this line is part of a multiline event, the next line will be part, too, put it into pending.
       event.tag(MULTILINE_TAG)
-      # this line is part of a multiline event, the next
-      # line will be part, too, put it into pending.
-      if pending
-        pending.append(event)
-      else
-        @pending[key] = event
-      end
+      pending << event
       event.cancel
     else
-      # if we have something in pending, join it with this message
-      # and send it. otherwise, this is a new message and not part of
-      # multiline, send it.
-      if pending
-        pending.append(event)
-        event.overwrite(pending)
-        @pending.delete(key)
+      # if we have something in pending, join it with this message and send it.
+      # otherwise, this is a new message and not part of multiline, send it.
+      unless pending.empty?
+        event.overwrite(merge(pending << event))
+        pending.clear
       end
     end # if match
   end
 
-  def collapse_event!(event)
-    event["message"] = event["message"].join("\n") if event["message"].is_a?(Array)
-    event.timestamp = event.timestamp.first if event.timestamp.is_a?(Array)
-    event
+  # merge a list of events. @timestamp for the resulting merged event will be from
+  # the "oldest" (events.first). all @source fields will be deduplicated depending
+  # on @allow_duplicates and joined with \n. all other fields will be deduplicated.
+  # @param events [Array<Event>] the list of events to merge
+  # @return [Event] the resulting merged event
+  def merge(events)
+    dups_key = @allow_duplicates ? @source : nil
+
+    data = events.inject({}) do |result, event|
+      self.class.event_hash_merge!(result, event.to_hash_with_metadata, dups_key)
+    end
+
+    # merged event @timestamp is from first event in sequence
+    data["@timestamp"] = Array(data["@timestamp"]).first
+    # collapse all @source field values
+    data[@source] = Array(data[@source]).join("\n")
+    LogStash::Event.new(data)
   end
+
+  # merge two events data hash, src into dst and handle duplicate values for dups_key
+  # @param dst [Hash] the event to merge into, dst will be mutated
+  # @param src [Hash] the event to merge in dst
+  # @param dups_key [String] the field key to keep duplicate values
+  # @return [Hash] mutated dst
+  def self.event_hash_merge!(dst, src, dups_key = nil)
+    src.each do |key, svalue|
+      dst[key] = if dst.has_key?(key)
+        dvalue = dst[key]
+
+        if dvalue.is_a?(Hash) && svalue.is_a?(Hash)
+          event_hash_merge!(dvalue, svalue, dups_key)
+        else
+          v = (dups_key == key) ? Array(dvalue) + Array(svalue) : Array(dvalue) | Array(svalue)
+          # the v result is always an Array, if none of the fields were arrays and there is a
+          # single value in the array, return the value, not the array
+          dvalue.is_a?(Array) || svalue.is_a?(Array) ? v : (v.size == 1 ? v.first : v)
```

Comment:
regarding the comment above, maybe

`dvalue.is_a?(Array) || svalue.is_a?(Array) ? v : (v.size == 1 ? v.first : v)``

should be

`dvalue.is_a?(Array) || svalue.is_a?(Array) ? v : v.first`

though I'm not sure the impact this has on other inputs